### ISSUE TITLE
Problem with `feature removed`

### ISSUE THREAD

**Issue by lethal233** [Jaic1/xv6-riscv-rust#1]
When I run the command in the repo
```shell
cargo +nightly run
```
It shows that
```shell
error[E0557]: feature has been removed
 --> src/lib.rs:4:12
  |
4 | #![feature(const_in_array_repeat_expressions)]
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ feature has been removed
  |
  = note: removed due to causing promotable bugs
```
And I comment on this line and `cargo run` again, it shows that 
```shell
#![no_std]
error[E0277]: the trait bound `PageTableEntry: core::marker::Copy` is not satisfied
  --> src/mm/pagetable.rs:76:19
   |
76 |             data: [PageTableEntry { data: 0 }; 512],
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `core::marker::Copy` is not implemented for `PageTableEntry`
   |
   = note: the `Copy` trait is required because the repeated element will be copied

error: aborting due to previous error; 1 warning emitted
```
Could you please tell me how to solve this problem?

Looking forward to your reply.

Lori

**Comment by lethal233** [comment]
Fine I add a comment in the line `#![feature(const_in_array_repeat_expressions)]`

and add `#[derive(Copy, Clone)]` in the file `src/mm/pagetable.rs` line:27 before the `pub struct PageTableEntry`.

and it runs as the result:

```shell
xv6-riscv-rust is booting

KernelHeap: available physical memory [0x8003b821, 0x88000000)
buddy system: useful memory is 0x7fc4000 bytes
buddy system: leaf size is 16 bytes
buddy system: free lists have 24 different sizes
buddy system: alloc 0x18004a0 bytes meta data
buddy system: 0x3c000 bytes unavailable
KernelHeap: init memory done
kvm_map: va=0x10000000, pa=0x10000000, size=0x1000
kvm_map: va=0x10001000, pa=0x10001000, size=0x1000
kvm_map: va=0x2000000, pa=0x2000000, size=0x10000
kvm_map: va=0xc000000, pa=0xc000000, size=0x400000
kvm_map: va=0x80000000, pa=0x80000000, size=0x18000
kvm_map: va=0x80018000, pa=0x80018000, size=0x7fe8000
kvm_map: va=0x3ffffff000, pa=0x80001000, size=0x1000
kvm_map: va=0x3fffffd000, pa=0x87fc1000, size=0x1000
kvm_map: va=0x3fffffb000, pa=0x87fc2000, size=0x1000
kvm_map: va=0x3fffff9000, pa=0x87fc3000, size=0x1000
kvm_map: va=0x3fffff7000, pa=0x87fc4000, size=0x1000
kvm_map: va=0x3fffff5000, pa=0x87fc5000, size=0x1000
kvm_map: va=0x3fffff3000, pa=0x87fc6000, size=0x1000
kvm_map: va=0x3fffff1000, pa=0x87fc7000, size=0x1000
kvm_map: va=0x3ffffef000, pa=0x87fc8000, size=0x1000
kvm_map: va=0x3ffffed000, pa=0x87fc9000, size=0x1000
kvm_map: va=0x3ffffeb000, pa=0x87fca000, size=0x1000
kvm_map: va=0x3ffffe9000, pa=0x87fcb000, size=0x1000
kvm_map: va=0x3ffffe7000, pa=0x87fcc000, size=0x1000
kvm_map: va=0x3ffffe5000, pa=0x87fcd000, size=0x1000
kvm_map: va=0x3ffffe3000, pa=0x87fce000, size=0x1000
kvm_map: va=0x3ffffe1000, pa=0x87fcf000, size=0x1000
kvm_map: va=0x3ffffdf000, pa=0x87fd0000, size=0x1000
kvm_map: va=0x3ffffdd000, pa=0x87fd1000, size=0x1000
kvm_map: va=0x3ffffdb000, pa=0x87fd2000, size=0x1000
kvm_map: va=0x3ffffd9000, pa=0x87fd3000, size=0x1000
kvm_map: va=0x3ffffd7000, pa=0x87fd4000, size=0x1000
kvm_map: va=0x3ffffd5000, pa=0x87fd5000, size=0x1000
kvm_map: va=0x3ffffd3000, pa=0x87fd6000, size=0x1000
kvm_map: va=0x3ffffd1000, pa=0x87fd7000, size=0x1000
kvm_map: va=0x3ffffcf000, pa=0x87fd8000, size=0x1000
kvm_map: va=0x3ffffcd000, pa=0x87fd9000, size=0x1000
kvm_map: va=0x3ffffcb000, pa=0x87fda000, size=0x1000
kvm_map: va=0x3ffffc9000, pa=0x87fdb000, size=0x1000
kvm_map: va=0x3ffffc7000, pa=0x87fdc000, size=0x1000
kvm_map: va=0x3ffffc5000, pa=0x87fdd000, size=0x1000
kvm_map: va=0x3ffffc3000, pa=0x87fde000, size=0x1000
kvm_map: va=0x3ffffc1000, pa=0x87fdf000, size=0x1000
kvm_map: va=0x3ffffbf000, pa=0x87fe0000, size=0x1000
kvm_map: va=0x3ffffbd000, pa=0x87fe1000, size=0x1000
kvm_map: va=0x3ffffbb000, pa=0x87fe2000, size=0x1000
kvm_map: va=0x3ffffb9000, pa=0x87fe3000, size=0x1000
kvm_map: va=0x3ffffb7000, pa=0x87fe4000, size=0x1000
kvm_map: va=0x3ffffb5000, pa=0x87fe5000, size=0x1000
kvm_map: va=0x3ffffb3000, pa=0x87fe6000, size=0x1000
kvm_map: va=0x3ffffb1000, pa=0x87fe7000, size=0x1000
kvm_map: va=0x3ffffaf000, pa=0x87fe8000, size=0x1000
kvm_map: va=0x3ffffad000, pa=0x87fe9000, size=0x1000
kvm_map: va=0x3ffffab000, pa=0x87fea000, size=0x1000
kvm_map: va=0x3ffffa9000, pa=0x87feb000, size=0x1000
kvm_map: va=0x3ffffa7000, pa=0x87fec000, size=0x1000
kvm_map: va=0x3ffffa5000, pa=0x87fed000, size=0x1000
kvm_map: va=0x3ffffa3000, pa=0x87fee000, size=0x1000
kvm_map: va=0x3ffffa1000, pa=0x87fef000, size=0x1000
kvm_map: va=0x3ffff9f000, pa=0x87ff0000, size=0x1000
kvm_map: va=0x3ffff9d000, pa=0x87ff1000, size=0x1000
kvm_map: va=0x3ffff9b000, pa=0x87ff2000, size=0x1000
kvm_map: va=0x3ffff99000, pa=0x87ff3000, size=0x1000
kvm_map: va=0x3ffff97000, pa=0x87ff4000, size=0x1000
kvm_map: va=0x3ffff95000, pa=0x87ff5000, size=0x1000
kvm_map: va=0x3ffff93000, pa=0x87ff6000, size=0x1000
kvm_map: va=0x3ffff91000, pa=0x87ff7000, size=0x1000
kvm_map: va=0x3ffff8f000, pa=0x87ff8000, size=0x1000
kvm_map: va=0x3ffff8d000, pa=0x87ff9000, size=0x1000
kvm_map: va=0x3ffff8b000, pa=0x87ffa000, size=0x1000
kvm_map: va=0x3ffff89000, pa=0x87ffb000, size=0x1000
kvm_map: va=0x3ffff87000, pa=0x8187c000, size=0x1000
kvm_map: va=0x3ffff85000, pa=0x8187d000, size=0x1000
kvm_map: va=0x3ffff83000, pa=0x8187e000, size=0x1000
kvm_map: va=0x3ffff81000, pa=0x8187f000, size=0x1000
kvm_map: va=0x3ffff7f000, pa=0x81880000, size=0x1000
hart 1 starting
hart 2 starting
read file system super block..done
sys_exec: /init
panicked at 'sys_exec: end', src/process/syscall.rs:46:9
```

Is the result right?

Looking forward to your reply.

Lori

**Comment by Jaic1** [comment]
> Fine I add a comment in the line `#![feature(const_in_array_repeat_expressions)]`
> 
> and add `#[derive(Copy, Clone)]` in the file `src/mm/pagetable.rs` line:27 before the `pub struct PageTableEntry`.
> 
> and it runs as the result:
> 
> ```shell
> xv6-riscv-rust is booting
> 
> KernelHeap: available physical memory [0x8003b821, 0x88000000)
> buddy system: useful memory is 0x7fc4000 bytes
> buddy system: leaf size is 16 bytes
> buddy system: free lists have 24 different sizes
> buddy system: alloc 0x18004a0 bytes meta data
> buddy system: 0x3c000 bytes unavailable
> KernelHeap: init memory done
> kvm_map: va=0x10000000, pa=0x10000000, size=0x1000
> kvm_map: va=0x10001000, pa=0x10001000, size=0x1000
> kvm_map: va=0x2000000, pa=0x2000000, size=0x10000
> kvm_map: va=0xc000000, pa=0xc000000, size=0x400000
> kvm_map: va=0x80000000, pa=0x80000000, size=0x18000
> kvm_map: va=0x80018000, pa=0x80018000, size=0x7fe8000
> kvm_map: va=0x3ffffff000, pa=0x80001000, size=0x1000
> kvm_map: va=0x3fffffd000, pa=0x87fc1000, size=0x1000
> kvm_map: va=0x3fffffb000, pa=0x87fc2000, size=0x1000
> kvm_map: va=0x3fffff9000, pa=0x87fc3000, size=0x1000
> kvm_map: va=0x3fffff7000, pa=0x87fc4000, size=0x1000
> kvm_map: va=0x3fffff5000, pa=0x87fc5000, size=0x1000
> kvm_map: va=0x3fffff3000, pa=0x87fc6000, size=0x1000
> kvm_map: va=0x3fffff1000, pa=0x87fc7000, size=0x1000
> kvm_map: va=0x3ffffef000, pa=0x87fc8000, size=0x1000
> kvm_map: va=0x3ffffed000, pa=0x87fc9000, size=0x1000
> kvm_map: va=0x3ffffeb000, pa=0x87fca000, size=0x1000
> kvm_map: va=0x3ffffe9000, pa=0x87fcb000, size=0x1000
> kvm_map: va=0x3ffffe7000, pa=0x87fcc000, size=0x1000
> kvm_map: va=0x3ffffe5000, pa=0x87fcd000, size=0x1000
> kvm_map: va=0x3ffffe3000, pa=0x87fce000, size=0x1000
> kvm_map: va=0x3ffffe1000, pa=0x87fcf000, size=0x1000
> kvm_map: va=0x3ffffdf000, pa=0x87fd0000, size=0x1000
> kvm_map: va=0x3ffffdd000, pa=0x87fd1000, size=0x1000
> kvm_map: va=0x3ffffdb000, pa=0x87fd2000, size=0x1000
> kvm_map: va=0x3ffffd9000, pa=0x87fd3000, size=0x1000
> kvm_map: va=0x3ffffd7000, pa=0x87fd4000, size=0x1000
> kvm_map: va=0x3ffffd5000, pa=0x87fd5000, size=0x1000
> kvm_map: va=0x3ffffd3000, pa=0x87fd6000, size=0x1000
> kvm_map: va=0x3ffffd1000, pa=0x87fd7000, size=0x1000
> kvm_map: va=0x3ffffcf000, pa=0x87fd8000, size=0x1000
> kvm_map: va=0x3ffffcd000, pa=0x87fd9000, size=0x1000
> kvm_map: va=0x3ffffcb000, pa=0x87fda000, size=0x1000
> kvm_map: va=0x3ffffc9000, pa=0x87fdb000, size=0x1000
> kvm_map: va=0x3ffffc7000, pa=0x87fdc000, size=0x1000
> kvm_map: va=0x3ffffc5000, pa=0x87fdd000, size=0x1000
> kvm_map: va=0x3ffffc3000, pa=0x87fde000, size=0x1000
> kvm_map: va=0x3ffffc1000, pa=0x87fdf000, size=0x1000
> kvm_map: va=0x3ffffbf000, pa=0x87fe0000, size=0x1000
> kvm_map: va=0x3ffffbd000, pa=0x87fe1000, size=0x1000
> kvm_map: va=0x3ffffbb000, pa=0x87fe2000, size=0x1000
> kvm_map: va=0x3ffffb9000, pa=0x87fe3000, size=0x1000
> kvm_map: va=0x3ffffb7000, pa=0x87fe4000, size=0x1000
> kvm_map: va=0x3ffffb5000, pa=0x87fe5000, size=0x1000
> kvm_map: va=0x3ffffb3000, pa=0x87fe6000, size=0x1000
> kvm_map: va=0x3ffffb1000, pa=0x87fe7000, size=0x1000
> kvm_map: va=0x3ffffaf000, pa=0x87fe8000, size=0x1000
> kvm_map: va=0x3ffffad000, pa=0x87fe9000, size=0x1000
> kvm_map: va=0x3ffffab000, pa=0x87fea000, size=0x1000
> kvm_map: va=0x3ffffa9000, pa=0x87feb000, size=0x1000
> kvm_map: va=0x3ffffa7000, pa=0x87fec000, size=0x1000
> kvm_map: va=0x3ffffa5000, pa=0x87fed000, size=0x1000
> kvm_map: va=0x3ffffa3000, pa=0x87fee000, size=0x1000
> kvm_map: va=0x3ffffa1000, pa=0x87fef000, size=0x1000
> kvm_map: va=0x3ffff9f000, pa=0x87ff0000, size=0x1000
> kvm_map: va=0x3ffff9d000, pa=0x87ff1000, size=0x1000
> kvm_map: va=0x3ffff9b000, pa=0x87ff2000, size=0x1000
> kvm_map: va=0x3ffff99000, pa=0x87ff3000, size=0x1000
> kvm_map: va=0x3ffff97000, pa=0x87ff4000, size=0x1000
> kvm_map: va=0x3ffff95000, pa=0x87ff5000, size=0x1000
> kvm_map: va=0x3ffff93000, pa=0x87ff6000, size=0x1000
> kvm_map: va=0x3ffff91000, pa=0x87ff7000, size=0x1000
> kvm_map: va=0x3ffff8f000, pa=0x87ff8000, size=0x1000
> kvm_map: va=0x3ffff8d000, pa=0x87ff9000, size=0x1000
> kvm_map: va=0x3ffff8b000, pa=0x87ffa000, size=0x1000
> kvm_map: va=0x3ffff89000, pa=0x87ffb000, size=0x1000
> kvm_map: va=0x3ffff87000, pa=0x8187c000, size=0x1000
> kvm_map: va=0x3ffff85000, pa=0x8187d000, size=0x1000
> kvm_map: va=0x3ffff83000, pa=0x8187e000, size=0x1000
> kvm_map: va=0x3ffff81000, pa=0x8187f000, size=0x1000
> kvm_map: va=0x3ffff7f000, pa=0x81880000, size=0x1000
> hart 1 starting
> hart 2 starting
> read file system super block..done
> sys_exec: /init
> panicked at 'sys_exec: end', src/process/syscall.rs:46:9
> ```
> 
> Is the result right?
> 
> Looking forward to your reply.
> 
> Lori

Your final result is expected.
I might fix this soon. :)

**Comment by Jaic1** [comment]
Fixed by 192aa204022d979b1b9af9aba0ccca066d7811b1.
